docs/source/architecture.md

```diff
@@ -95,16 +95,41 @@ so less restrictions apply to this than to `get_build_scripts`.
 
 At the end of the assemble step, the docker image is ready to be used in various ways!
 
-## Push
+### Push
 
 Optionally, repo2docker can **push** a built image to a [docker registry](https://docs.docker.com/registry/).
 This is done as a convenience only (since you can do the same with a `docker push` after using repo2docker
 only to build), and implemented in `Repo2Docker.push` method. It is only activated if using the
 `--push` commandline flag.
 
-## Run
+### Run
 
 Optionally, repo2docker can **run** the built image and allow the user to access the Jupyter Notebook
 running inside by default. This is also done as a convenience only (since you can do the same with `docker run`
 after using repo2docker only to build), and implemented in `Repo2Docker.run`. It is activated by default
-unless the `--no-run` commandline flag is passed.
\ No newline at end of file
+unless the `--no-run` commandline flag is passed.
+
+## ContentProviders
+
+ContentProviders provide a way for `repo2docker` to know how to find and
+retrieve a repository. They follow a similar pattern as the BuildPacks
+described above. When `repo2docker` is called, its main argument will be
+a path to a repository. This might be a local path or a URL. Upon being called,
+`repo2docker` will loop through all ContentProviders and perform the following
+commands:
+
+* Run the `detect()` method on the repository path given to `repo2docker`. This
+  should return `True` if the path matches what the ContentProvider is looking
+  for.
+
+  > For example, the [`Local` ContentProvider](https://github.com/jupyter/repo2docker/blob/80b979f8580ddef184d2ba7d354e7a833cfa38a4/repo2docker/contentproviders/base.py#L64)
+  > checks whether the argument is a valid local path. If so, then `detect(`
+  > returns true.
```

Comment:
```suggestion
  > returns True.
```

is this correct?

Reply:
No. All content providers should (and do) return `None` if they don't match and something "that isn't `None`" when they do match. What precisely it is they return is left to them. The only requirement is that the `fetch()` method of the provider knows what to do with it.

Reply:
I was asking in this case specifically about the `Local` contentprovider rather than in general. I'll dig into it a bit more to figure it out

Reply:
see updated edit above, LMK if that's OK

Reply:
> was asking in this case specifically

Ah ok. I missed that. 

Yes the suggestion is correct.

Reply:
Do you have any idea why I can't commit the suggestion? Normally it works but it seems for this PR it consistently doesn't (and there are other PRs where it doesn't). So confusing.